### ISSUE TITLE
NaN on Summit with offload code

### ISSUE THREAD

**Issue by markdewing** [QMCPACK/qmcpack#2864]
Some optimizer runs will trip the NaN checks in QMCHamiltonian.cpp.   It happens during the batched VMC while generating samples.   The NaN's were traced back to psiMinv in DiractDeterminantBatched::updateAfterSweep.

The parameters are frozen, so it is not an issue of parameters changing between optimizer loop iterations.

**To Reproduce**
The input file was modified from the NiO performance tests, size dmc-a32-e384-batched_driver-DU32
The problem occurs with delay rank 32 (and 8).  It does not occur with delay rank 1 or 2.
There are 128 walkers per rank

It depends on a particular random number seed.  Some runs show no problems. (So clearly there is some dependence on the path taken through the program)

**System:**
Summit.  Using XL compiler

Compiled with offload and ENABLE_CUDA=1.  (If ENABLE_CUDA is turned off, the problem does not happen.)

Launch command is: `jsrun -n 1 -g 1 -c 7 -b packed:7 ../../offload_qmcpack --enable-timers=medium opt.xml`

```
Currently Loaded Modules:
  1) hsi/5.0.2.p5    4) darshan-runtime/3.1.7   7) spectrum-mpi/10.3.1.2-20200121  10) cuda/10.1.243        13) hdf5/1.10.4
  2) xalt/1.2.1      5) DefApps                 8) cmake/3.18.2                    11) essl/6.1.0-2         14) python/3.6.6-anaconda3-5.3.0
  3) lsf-tools/2.0   6) xl/16.1.1-5             9) git/2.20.1                      12) netlib-lapack/3.8.0  15) libunwind/1.2.1
```

**Additional context**

Code from DiracDeterminantBatched.cpp where the problem seems to be happening.
```
​template<typename DET_ENGINE_TYPE>
void DiracDeterminantBatched<DET_ENGINE_TYPE>::mw_completeUpdates(const RefVector<WaveFunctionComponent>& WFC_list)
{
  RefVector<DET_ENGINE_TYPE> engine_list;
  engine_list.reserve(WFC_list.size());
  for (int iw = 0; iw < WFC_list.size(); iw++)
  {
    auto& det = static_cast<DiracDeterminantBatched<DET_ENGINE_TYPE>&>(WFC_list[iw].get());
    engine_list.push_back(det.det_engine_);
  }
  //
  // If this block is enabled and compiled, the problem goes away
  //
#if 0
    for (int iw = 0; iw < engine_list.size(); iw++) {
      auto& det = engine_list[iw].get();
      if (std::isnan(det.psiMinv[0][0])) {
          app_log() << "DiracDeterminantBatched::mw_CompleteUpdates before psiMinv is NaN" << std::endl;
          app_log() << "  psiMinv[0][0] = " << det.psiMinv[0][0] << std::endl;
          APP_ABORT("NaN");
      }
    }
#endif
{
    ScopedTimer update(&UpdateTimer);
    det_engine_.mw_updateInvMat(engine_list);
  }

  {
    ScopedTimer d2h(&D2HTimer);

    for (int iw = 0; iw < WFC_list.size(); iw++)
    {
      auto& det          = static_cast<DiracDeterminantBatched<DET_ENGINE_TYPE>&>(WFC_list[iw].get());
      auto& my_psiM_vgl  = det.psiM_vgl;
      auto* psiM_vgl_ptr = my_psiM_vgl.data();
      PRAGMA_OFFLOAD("omp target update from(psiM_vgl_ptr[my_psiM_vgl.capacity():my_psiM_vgl.capacity()*4]) nowait")
    }

    det_engine_.mw_transferAinv_D2H(engine_list);

    PRAGMA_OFFLOAD("omp taskwait")



    for (int iw = 0; iw < engine_list.size(); iw++) {
      auto& det = engine_list[iw].get();
      // 
      //  When it goes bad, here is where the the NaN is caught
     //
      if (std::isnan(det.psiMinv[0][0])) {
          app_log() << "DiracDeterminantBatched::mw_CompleteUpdates after D2H psiMinv is NaN" << std::endl;
          app_log() << "  psiMinv[0][0] = " << det.psiMinv[0][0] << std::endl;
          APP_ABORT("NaN");
      }
    }
  }
}
```

**Comment by PDoakORNL** [comment]
very much reminds me of the experience I had in DiracDeterminant.cpp that led to addition of
```c++
assert(WorkingIndex >= 0);
```
several places.

I ended up with a believe a corrupt psiMinv or invRow because of the strong assumption in DiracDeterminant has that the caller would never call 
```c++
DiracDeterminant this_determinant;
...
'this_determinant.evalXXX(...,iat,...)'
with an `iat < this_determinant.FirstIndex`

In debug this causes an access violation (sometimes) but in release just NaN.

**Comment by markdewing** [comment]
The issue can be reproduced with just VMC.  I changed qmc method to 'vmc_batch' and unrolled the loop. (Using vmc_batch inside a loop doesn't work - it segfaults on the RNG.  It has to do with the driver lifetime inside the loop. Not important for now.)

I tried collapsing the four VMC input blocks to a single one with more blocks, but that did not show the issue, and ran okay.

Different runs reliably reproduce, but the NaN's occur on a different walker.   With some of the optimization cases, the VMC energy was different between runs.

There is a comment in the code segment in the original description that mentions the problem goes away if a particular section is included.   That only happens with the optimizer input file - it does not happen with this VMC input file (that is, including the code and the NaN's still occur.)

My current theory is there some sequence of delayed matrix updates that causes access to uninitialized memory, or out-of-bound accesses. 

I tried running cuda-memcheck, but it ran out of time before finishing.  There were about 40 minutes left on the interactive session.  I should try again with a 2 hour session on the debug queue.



[vmc_input.zip](https://github.com/QMCPACK/qmcpack/files/5899225/vmc_input.zip)

**Comment by prckent** [comment]
Thanks for pursuing this Mark. So the delayed update code has a bug. Does this only occur with OpenMP offload or have you ever seen it with pure CPU runs?

**Comment by markdewing** [comment]
I have not seen it with CPU runs.   It only occurs with OMP offload + the CUDA extras.  Pure OMP offload doesn't have delayed update implemented, so the situation can't be recreated there.

The NaN's occur after one or two moves into the second VMC section.  It could be something with ending/starting a new section?

**Comment by markdewing** [comment]
Configuration is
```
cmake \
-D CMAKE_BUILD_TYPE=RelWithDebInfo \
-D CMAKE_C_COMPILER="mpicc" \
-D CMAKE_CXX_COMPILER="mpic++" \
-D BUILD_LMYENGINE_INTERFACE=0 \
-D QMC_DATA=/ccs/home/mdewing/qmcpack/test \
-D QMC_MPI=1 \
-D QMC_CUDA=0 \
-D QMC_COMPLEX=0 \
-D QMC_MIXED_PRECISION=0 \
-D BUILD_AFQMC=0 \
-D ENABLE_CUDA=1 \
-D CUDA_ARCH="sm_70" \
-D ENABLE_MASS=1 \
-D MASS_ROOT=/sw/summit/xl/16.1.1-5/xlmass/9.1.1 \
-D MPIEXEC_EXECUTABLE=`which jsrun` \
-D MPIEXEC_NUMPROC_FLAG='-n' \
-D MPIEXEC_PREFLAGS='-c;16;-g;1;-b;packed:16' \
-D CMAKE_C_FLAGS=-qarch=pwr9 \
-D CMAKE_CXX_FLAGS="-qarch=pwr9 -qxflag=disable__cplusplusOverride -isystem /sw/summit/gcc/6.4.0/include/c++/6.4.0/powerpc64le-none-linux-gnu -qgcc_cpp_stdinc=/sw/summit/gcc/6.4.0/include/c++/6.4.0" \
-D CMAKE_CXX_STANDARD_LIBRARIES=/sw/summit/gcc/6.4.0/lib64/libstdc++.a \
-D BLAS_essl_LIBRARY=$OLCF_ESSL_ROOT/lib64/libessl.so  \
-D ENABLE_OFFLOAD=1 \
-D ENABLE_TIMERS=1 \
-D USE_NVTX_API=1 \
-D CMAKE_EXPORT_COMPILE_COMMANDS=ON \
../qmcpack
```

The NaN's also occur in mixed precision.

Getting the USE_NVTX_API switch to work requires modifying the CMakelists.txt file.  (The CMake documentation on CMAKE_DEPENDENT_OPTIONS is not consistent with how the macro actually works - complex conditionals do not work.  I need to file a CMake issue on this).  Simplify the conditional by removing QMC_CUDA, and then it works
```
CMAKE_DEPENDENT_OPTION(USE_NVTX_API "Enable/disable NVTX regions in CUDA code." OFF "ENABLE_TIMERS AND ENABLE_CUDA" OFF)
```

**Comment by ye-luo** [comment]
@markdewing could you try on your side to see if the current develop still ends up with NaNs? I tried on my side and full/mixed precision real build are just fine with XL build on Summit.

**Comment by markdewing** [comment]
I updated to the latest version, and it does not reproduce the NaN issue.  I will try previous versions to see if it reproduces.    I checked out a fresh clone of the qmcpack source, and compared with what I was using, and nothing in the diffs stood out as a possible cause or resolution to the issue.

**Comment by ye-luo** [comment]
@markdewing FYI #2849 and #2857 are potential relavent fixes.

**Comment by markdewing** [comment]
So #2849 is not it - the previous version I used included that one.   It might be #2857 - a build from the version where that PR is merged works.   I'm  compiling one revision prior to check.

**Comment by markdewing** [comment]
The NaN problem can be reproduced with this revision (the merge for #2586), the one prior to merging #2857 : 
 c3479b49315530b40bef7eb34e6d70fcf1a5370f

**Comment by ye-luo** [comment]
Thanks @markdewing for narrow this down to a commit.
Although I still don't know why #2857 fixes the problem as I mentioned in the PR, at least everything runs fine right now. Let us close this right now unless the same issue reappears.

**Comment by markdewing** [comment]
The VMC example works, but the optimizer example still gets NaN in the gradient for the VMC drift moves.

**Comment by ye-luo** [comment]
@markdewing try out the develop, the NaN issue should have been fixed by #2876.

**Comment by markdewing** [comment]
Seems to be fixed - there are no NaN's when running the the optimizer example
